### PR TITLE
Changes to Metrics, Augmentations, Configuration, and Model Factory

### PR DESCRIPTION
**Pull request by jansky**

### CODE REVIEW

**Comment by jansky** [comment]
I have made the following updates:

## Discolight

Discolight augmentations are now supported.

## Metrics

A new `metrics.py` file defines a 100%-configurable system of computing model validation metrics. Individual metrics are implemented as classes, and may use the values of previously-computed metrics in their own calculations. Automated dependency injection is used to ensure the correct computation order and delivery of needed metric values. The `metrics.py` file is thoroughly commented to describe these changes, but don't hesitate to let me know if you have questions. This file contains the most complex changes.

The metrics you want to use are now specified in the `config.yaml` file. Certain metrics extend the `ComparableMetric` class, which means the `Trainer` will keep track of their best value as training and evaluation progresses. The configuration field `monitored_metric` should be set to the name of the metric that determines when new weights are saved. When a new best value for this metric is found after validation, the model weights are saved.

## Automated Configuration Generation

For the `optimizer`, `criterion`, `criterion_val`, `scheduler`, and `metrics` configuration fields, specifying an object without a configuration will cause the `YAMLConfig` class to generate a default configuration for the object and save it to the specified configuration file. These changes are implemented in `config.py`.

## Configurable Model Factory

The `CustomEfficientNet` class can be configured to use `geffnet` or `timm` to load the specified model by setting the `model_factory` field in `config.yaml`. Custom versions of `geffnet` and `timm` are supplied that support setting the `model_weight_path_folder` configuration option to change where downloaded pretrained models are loaded and stored.
